src/main.rs

```diff
@@ -0,0 +1,90 @@
+use crate::downloader::Downloader;
+use crate::manifest::Manifest;
+use crate::storage::{CdnReader, FileStatus, S3Storage, Storage};
+use anyhow::Error;
+use clap::Parser;
+use std::path::PathBuf;
+
+mod downloader;
+mod manifest;
+mod storage;
+
+#[derive(Debug, Parser)]
+struct Cli {
+    /// Path to the manifest to synchronize.
+    #[arg(default_value = "files.toml")]
+    manifest: PathBuf,
+
+    /// Only check which changes are needed (no credentials required).
+    #[arg(long)]
+    skip_upload: bool,
+
+    /// Base URL of the CDN where mirrored files are served.
+    #[arg(long, default_value = "https://ci-mirrors.rust-lang.org")]
+    cdn_url: String,
+
+    /// Name of the S3 bucket containing the files.
+    #[arg(long, default_value = "rust-lang-ci-mirrors")]
+    s3_bucket: String,
+}
+
+fn main() -> Result<(), Error> {
+    let args = Cli::parse();
+    let manifest = Manifest::load(&args.manifest)?;
+
+    let storage = if args.skip_upload {
+        Storage::ReadOnly(CdnReader::new(args.cdn_url))
+    } else {
+        Storage::ReadWrite(S3Storage::new(args.s3_bucket)?)
+    };
+
+    // Collect all errors that happen during the check phase and show them at the end. This way, if
+    // there are multiple errors in CI users won't have to retry the build multiple times.
+    let mut errors = Vec::new();
+
+    eprintln!("calculating the changes to execute...");
+    let mut to_upload = Vec::new();
+    for file in &manifest.files {
+        let name = &file.name;
+        match storage.file_status(&file.name)? {
+            FileStatus::Legacy => errors.push(format!(
+                "file {name} was already uploaded without this tool"
+            )),
+            FileStatus::Present { sha256 } if sha256 != file.sha256 => errors.push(format!(
+                "file {name} was already uploaded with different content"
+            )),
+            FileStatus::Missing => to_upload.push(file),
+            FileStatus::Present { .. } => {}
+        }
+    }
+
+    // We download eagerly to be able to detect errors during the check phase.
+    let downloader = Downloader::new()?;
```

Comment:
Probably fine for now, but the eager download could cost quite a bit of disk space. Could be annoying when running locally, and eventually could even run out of space on CI.

Reply:
Right, of course, sorry. That's what I get for reviewing a PR on my phone :facepalm: